### ISSUE TITLE
Memoization (release 2.9.8) is not always good

### ISSUE THREAD

**Issue by b3b00** [b3b00/csly#435]
parser memoization introduced with release 2.9.8 (fix #414) is not always faster as benchcurrent shows (on branch [performance/memoization-slows-down](https://github.com/b3b00/csly/tree/performance/memoization-slows-down)) : 

```

BenchmarkDotNet v0.13.12, Windows 10 (10.0.19045.4291/22H2/2022Update)
Intel Core i7-10610U CPU 1.80GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.104
  [Host]     : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2


```
| Method   | Memoize | Mean      | Error    | StdDev   | Gen0       | Gen1      | Gen2      | Allocated |
|--------- |-------- |----------:|---------:|---------:|-----------:|----------:|----------:|----------:|
| **TestJson** | **False**   |  **88.09 ms** | **1.714 ms** | **1.905 ms** |  **9500.0000** | **3500.0000** | **1500.0000** |     **54 MB** |
| **TestJson** | **True**    | **122.42 ms** | **2.438 ms** | **4.925 ms** | **10800.0000** | **3800.0000** | **1800.0000** |  **66.62 MB** |

Memoization works well when a grammar leads to heavy backtracking. In the case of JSON backtracing is limited and memoization slow down the parse.
Memoization should be a opt-in.

Add another benchmark that could demonstrate the added value  of memoization in case of heavy backtracking (see #414 )

**Comment by b3b00** [comment]
using #414 parser as a base memoization shows its value : 

```

BenchmarkDotNet v0.13.12, Windows 10 (10.0.19045.4291/22H2/2022Update)
Intel Core i7-10610U CPU 1.80GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.104
  [Host]     : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.4 (7.0.423.11508), X64 RyuJIT AVX2


```
| Method        | Memoize | Mean      | Error     | StdDev    | Gen0        | Gen1       | Gen2      | Allocated |
|-------------- |-------- |----------:|----------:|----------:|------------:|-----------:|----------:|----------:|
| **TestBackTrack** | **False**   | **809.85 ms** | **15.010 ms** | **34.184 ms** | **169000.0000** | **39000.0000** | **8000.0000** |  **929.6 MB** |
| **TestBackTrack** | **True**    |  **33.83 ms** |  **1.289 ms** |  **3.697 ms** |   **1812.5000** |  **1781.2500** | **1406.2500** |   **23.8 MB** |

**Comment by b3b00** [comment]
Fixed #436
